src/graphene.cpp

```diff
@@ -0,0 +1,1548 @@
+    bool fXVal;
+    {
+        LOCK(cs_main);
+        fXVal = (header.hashPrevBlock == chainActive.Tip()->GetBlockHash()) ? true : false;
+    }
+
+    graphenedata.ClearGrapheneBlockData(pfrom);
+    pfrom->nSizeGrapheneBlock = nSizeGrapheneBlock;
+
+    uint256 nullhash;
+    pfrom->grapheneBlock.nVersion = header.nVersion;
+    pfrom->grapheneBlock.nBits = header.nBits;
+    pfrom->grapheneBlock.nNonce = header.nNonce;
+    pfrom->grapheneBlock.nTime = header.nTime;
+    pfrom->grapheneBlock.hashMerkleRoot = header.hashMerkleRoot;
+    pfrom->grapheneBlock.hashPrevBlock = header.hashPrevBlock;
+    pfrom->grapheneBlockHashes.clear();
+    pfrom->grapheneBlockHashes.resize(nBlockTxs, nullhash);
+
+    {
+        LOCK(pfrom->cs_grapheneadditionaltxs);
+
+        pfrom->grapheneAdditionalTxs.clear();
+        for (auto tx : vAdditionalTxs)
+            pfrom->grapheneAdditionalTxs.push_back(tx);
+    }
+
+    vTxHashes.reserve(nBlockTxs);
+
+    // Create a map of all 8 bytes tx hashes pointing to their full tx hash counterpart
+    // We need to check all transaction sources (orphan list, mempool, and new (incoming) transactions in this block)
+    // for a collision.
+    int missingCount = 0;
+    int unnecessaryCount = 0;
+    bool collision = false;
+    std::set<uint256> passingTxHashes;
+    std::map<uint64_t, uint256> mapPartialTxHash;
+    std::vector<uint256> memPoolHashes;
+    std::set<uint64_t> setHashesToRequest;
+
+    bool fMerkleRootCorrect = true;
+    {
+        // Do the orphans first before taking the mempool.cs lock, so that we maintain correct locking order.
+        LOCK(orphanpool.cs);
+        for (auto &kv : orphanpool.mapOrphanTransactions)
+        {
+            uint256 hash = kv.first;
+
+            uint64_t cheapHash = hash.GetCheapHash();
+
+            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
+                collision = true;
+
+            mapPartialTxHash[cheapHash] = hash;
+        }
+
+        // We don't have to keep the lock on mempool.cs here to do mempool.queryHashes
+        // but we take the lock anyway so we don't have to re-lock again later.
+        ////////////////////// What is cs_xval for?
+        LOCK(cs_xval);
+        mempool.queryHashes(memPoolHashes);
+
+        for (const uint256 &hash : memPoolHashes)
+        {
+            uint64_t cheapHash = hash.GetCheapHash();
+
+            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
+                collision = true;
+
+            mapPartialTxHash[cheapHash] = hash;
+        }
+
+        if (!collision)
+        {
+            std::vector<uint256> localHashes;
+            for (const std::pair<uint64_t, uint256> &kv : mapPartialTxHash)
+                localHashes.push_back(kv.second);
+
+            // Add full transactions included in the block
+            for (auto tx : vAdditionalTxs)
+                localHashes.push_back(tx->GetHash());
```

Comment:
use a reference for better performance so we don't have to do a copy

-for (auto tx : vAdditionalTxs)
+for (auto &tx : vAdditionalTxs)